# Машинное обучение. ВМК МГУ

# Практическое задание 2: Numpy + pandas + matplotlib

## Уровень: **Базовый (Base)**

# О формате сдачи

🔷 При решении ноутбука используйте данный шаблон

```
✅ Можно добавлять новые ячейки любых типов
❌ Не нужно удалять текстовые ячейки c разметкой частей ноутбука и формулировками заданий
```

🔷 При оценивании задач учитывается код

```
✅ Задания, в которых необходим код, обычно помечаются фразами "Your code here"/"Ваш код" и аналогичными
❌ Ответы на вопросы без сопутствующего кода оцениваются в 0 баллов
❌ Наличе работоспособного кода в ноутбуке, если на сказано иного, обязательно
```

🔷 При оценивании задач учитываются выводы

```
✅ Задания, в которых необходимы выводы, обычно помечаются фразами Вывод"/"Ответ на вопрос"/"Ваш текст" и аналогичными
✅ Обычно выводы подразумевают под собой текстовый ответ (можно писать markdown, latex).
✅ Сопутствующие изображения, графики, таблички - приветствуются!
❌ При отсутствии выводов задание не засчитается на полный балл
```

- Многие из заданий можно выполнить несколькими способами.
Не существуют единственно верного, но попробуйте максимально
задействовать арсенал pandas и ориентируйтесь на простоту и понятность
вашего кода. Не забывайте, что можно гуглить и что-то искать на stackoverflow, например.

# Подготовка рабочей среды

Сначала установим нужные нам версии библиотек. Мы гарантируем, что в данных версиях задание будет корректно отрабатывать.

После установки нужных версий, возможно, нужно перезагрузить среду (runtime), но скорее всего вам это не понадобится

На скачивание файла и установку понадобится не более 5 минут.

**Важно!**

Устанавливать нужные версии нужно каждый раз, когда создается новый
рантайм. Например, если вы 2 часа подряд делаете это задание, то
подготовить библиотеки достаточно 1 раз. Но если вы, например, начали в
понедельник, затем закрыли/выключили ноутбук, то при продолжении в
среду, вам нужно будет запустить рантайм заново и следовательно заново
установить библиотеки.

**Важно!**
Если вы предпочитаете делать практические задания на своем личном ноутбуке, то проверьте, что вы установили рабочее окружение в соответствии с гайдом .pdf)

In [ ]:
# !!! Данный блок будет работать только в Google-Colab !!!
! gdown 10k8Hwn9kpK9SpK4IEj4-EaWQZqgYT5-Q
! pip install -r /content/requirements_2024_25_for_colab_small.txt

Downloading...
From: https://drive.google.com/uc?id=10k8Hwn9kpK9SpK4IEj4-EaWQZqgYT5-Q
To: /content/requirements_2024_25_for_colab_small.txt
100% 375/375 [00:00<00:00, 1.16MB/s]
  Using cached jedi-0.19.1-py2.py3-none-any.whl.metadata (22 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 516.3/516.3 kB 31.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 83.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 86.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.5/41.5 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.5/15.5 MB 74.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.7/14.7 MB 78.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 79.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.8/294.8 kB 21.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.

Проверим версию библиотеки:

In [ ]:
import catboost
assert(catboost.__version__ == '1.2.7')

Теперь можно приступать к выполнению задания! :)

# О задании

Задание состоит из трех частей, посвященных работе с табличными данными с помощью библиотеки **pandas** , визуализации с помощью библиотек **matplotlib** , **seaborn** , **plotly** и векторным вычислениям с помощью библиотеки **numpy** . Вам предлагается выполнить несколько заданий.

# Часть 1. Pandas [5 баллов], кросс-проверка

В этой части вы отработаете навыки работы с библиотекой **pandas** .

**Подсказка** Если вы еще не прошли выданный воркбук по pandas , то советуем в начале обратиться к нему

Далее потребуется выполнить несколько небольших заданий, покрывающих базовый

1. Новый пункт
2. Новый пункт

функционал библиотеки, который довольно часто пригождается на первичных этапах анализа данных перед обучением ML-моделей.

In [ ]:
import pandas as pd

In [ ]:
!gdown 1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6

Downloading...
From: https://drive.google.com/uc?id=1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6
To: /content/youtube_trending_mediterranean.csv
100% 6.22M/6.22M [00:00<00:00, 76.5MB/s]


### **Задание 1 [0.25 баллa]**

Откройте файл с таблицей, загруженной выше, и выведите последние 5 строк.

**Подсказка** Полезные материалы: https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/02_read_write.html

Возможно, пригодятся такие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.tail.html#pandas.DataFrame.tail

In [ ]:
df = pd.read_csv("youtube_trending_mediterranean.csv")
df.tail(5)

,country,video_id,title,publishedAt,channelId,channelTitle,categoryId,trending_date,tags,view_count,comment_count,thumbnail_link,comments_disabled,ratings_disabled,description
3216,NaN,eB-A1Y0Dswc,Thank you for watching,2022-07-05T21:18:33Z,UCS5tt2z_DFvG7-39J3aE-bQ,Life of Boris,22,22.11.07,lifeofboris|life of boris|english subtitles|ye...,864216,14431,https://i.ytimg.com/vi/eB-A1Y0Dswc/default.jpg,False,True,New channel: https://www.youtube.com/channel/U...
3217,NaN,KJGTtzccOHU,Pokemon Emerald but every battle is extremely ...,2022-07-05T01:55:10Z,UC0VVYtw21rg2cokUystu2Dw,SmallAnt,20,22.11.07,pokemon|smallant|smallant1|challenge|pokemon c...,1311698,1232,https://i.ytimg.com/vi/KJGTtzccOHU/default.jpg,False,True,Pokemon emerald but every battle is a 1v2You s...
3218,NaN,be8mqaTi_DI,Victor Cibrian x Fuerza Regida x Luis R Conriq...,2022-07-04T21:00:36Z,UC648rgJOboZlgcDbW00vTSA,Rancho Humilde,10,22.11.07,Rancho Humilde|Legado 7|El De La Guitarra|Fuer...,2256933,597,https://i.ytimg.com/vi/be8mqaTi_DI/default.jpg,False,True,Suscríbete: https://Rancho.lnk.to/RHYoutube\r...
3219,NaN,14aYHc7ybu0,WE STAYED AT A 1 STAR HOTEL FOR 24 HOURS | coc...,2022-07-05T13:40:04Z,UCSGoIq_tVESqNYF1Re-zn1Q,HJ Evelyn,24,22.11.07,hjevelyn|Evelyn ha|Emily ha|erica ha|korean si...,922668,2951,https://i.ytimg.com/vi/14aYHc7ybu0/default.jpg,False,True,never again ahaha.... CHECK OUT OUR MERCH: htt...
3220,NaN,mQ9eFI13NCQ,"I 100%'d Cuphead: The Delicious Last Course, H...",2022-07-04T16:04:33Z,UCRgzckrm1Svf-Y_dIWsTPTg,The Andrew Collette Show,20,22.11.07,cuphead|cuphead dlc|cuphead the delicious last...,1176289,620,https://i.ytimg.com/vi/mQ9eFI13NCQ/default.jpg,False,True,My Cuphead: The Delicious Last Course Let's Pl...


### **Задание 2 [0.25 баллa]**

Сколько в таблице строк и столбцов?

**Подсказка**

Возможно, пригодятся такие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html#pandas.DataFrame.shape

In [ ]:
df.shape[0] #строк

3221


In [ ]:
df.shape[1] #столбцов

15


**3221 строк и 15 столбцов**

### **Задание 3 [0.5 баллa]**

Посмотрите на статистики значений в каждой колонке, отдельно для числовых и для не-числовых признаков.

Ответьте на вопросы:

- Чему равно стандартное отклонение (std) у колонки view_count ?
- Чем равно самое частое (топовое) значение колонки channelTitle ?

**Подсказка**

Возможно, пригодятся такие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html

In [ ]:
df["view_count"].std()
df['channelTitle'].describe()

,channelTitle
count,3221
unique,1344
top,Wimbledon
freq,25


**14008445.725553624, топ channelTitle Wimblendon**

### **Задание 4 [0.5 баллa]**

Ответьте на вопрос: как вы думаете, что это за данные, о чем они?

**Подсказка** Возможно вам понадобится посмотреть на таблицу разными методами, не стесняйтесь этого делать

In [ ]:
df.columns

Index(['country', 'video_id', 'title', 'publishedAt', 'channelId',
       'channelTitle', 'categoryId', 'trending_date', 'tags', 'view_count',
       'comment_count', 'thumbnail_link', 'comments_disabled',
       'ratings_disabled', 'description'],
      dtype='object')


**Судя по названиям столбцов, в табличке
представлены различные сведения о видео, в различное время выходивших в
тренды на ютубе**

### **Задание 5 [0.5 баллa]**

Для скольких стран (колонка country) есть информация о видеороликах в таблице (отсутствие страны НЕ учитывается как +1 значение) ?

**Подсказка**

Базовый функционал, необxодимый для решения, можно найти здесь .

Возможно, пригодятся следующие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.nunique.html#pandas.Series.nunique

In [ ]:
df['country'].nunique(True)

19


**19**

### **Задание 6 [0.5 баллa]**

Сколько
есть видеороликов (== записей в таблице) c разрешенными и запрещенными
комментариями? (колонка comments_disabled). Под видеороликом считаем 1
запись в таблице.

**Подсказка**

Возможно, пригодятся следующие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html#pandas.Series.value_counts

In [ ]:
df[df['comments_disabled'].isin([True])].shape[0]

28


In [ ]:
df[df['comments_disabled'].isin([False])].shape[0]

3193


**У 28 выключены и у 3193 включены**

### **Задание 7 [0.5 баллa]**

Какое максимальное число просмотров у видеороликов, попавших в таблицу? (колонка view_count)

**Подсказка**

Базовый функционал, необxодимый для решения, можно найти здесь .

Возможно, пригодятся следующие функции/методы:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.max.html#pandas.Series.max

In [ ]:
df['view_count'].max()

123034293


**123034293 просмотров**

### **Задание 8 [0.5 баллa]**

Есть ли в данных пропуски? В каких колонках?

**Подсказка**

Про работу с пропущенными значениями можно прочитать здесь .

Возможно, пригодится такая функция: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html#pandas.DataFrame.isnull

In [ ]:
selected_columns = ['country', 'video_id', 'title', 'publishedAt', 'channelId',
       'channelTitle', 'categoryId', 'trending_date', 'tags', 'view_count',
       'comment_count', 'thumbnail_link', 'comments_disabled',
       'ratings_disabled', 'description']
df[selected_columns].isna().any()

,0
country,True
video_id,False
title,False
publishedAt,False
channelId,False
channelTitle,False
categoryId,False
trending_date,False
tags,False
view_count,False


**country, description**

### **Задание 9 [0.5 баллa]**

Заполните пропуски пустой строкой для текстовых колонок, и удалите строки с пропусками для колонок с категориальными значениями.

**Подсказка**

Значения **категориальной**
переменной определяют факт принадлежности к какой-то категории,
категорий при этом должно быть ограниченное кол-во сильно меньшее
размера датасета.

**Категориальная переменная** -
колонка, которая имеет ограниченное число уникальных значений. Вообще
говоря, может быть как типа object (текстовые), так и int (численные). В
случае int можно считать, что категории закодированы числами. В таком
случае значения нельзя сравнивать между собой и совершать над ними
арифметические операции. В этом задании категориальные переменные только
типа object.

**Внимание!** Для проверки
корректности выведите кол-во пропусков во всех столбцах и общее кол-во
строк в таблице - должна быть 3021 запись. Ошибка при удалении строк
может привести к неправильному выполнению следующих заданий.

Возможно, пригодятся такие функции:

- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html#pandas.DataFrame.fillna
- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html#pandas.DataFrame.dropna

In [ ]:
df['description'] = df['description'].fillna('')
df = df.dropna(subset=['country'])

In [ ]:
assert df.shape[0] == 3021, "Проверьте правильность выполнения задания."

### **Задание 10 [0.5 баллa]**

Какие
5 роликов набрали в Италии (country='IT') наибольшее число просмотров
(колонка view_count)? В качестве ответа привидите названия каналов с
соответствующими роликами (колонка channelTitle) в порядке убывания
числа просмотров полученных роликов.

**Подсказка**

Для решения воспользуйтесь функцией sort_values .

In [ ]:
print(*[i for i in df[df['country'] == 'IT'].sort_values(by='view_count', ascending=False).head(5)['channelTitle']], sep=', ')

That Little Puff, SMTOWN, Celine & Michiel, SportsNation, BeatboxJCOP


**'That Little Puff, SMTOWN, Celine & Michiel, SportsNation, BeatboxJCOP**

### **Задание 11 [0.5 баллa]**

Какой видеоролик был опубликован в наибольшем кол-ве стран? В качестве ответа привeдите название видеоролика (колонка title).

**Подсказка**

Для решения могут помочь функции groupby и sort_values .

In [ ]:
df.groupby('title')['country'].nunique().sort_values(ascending=False).head(1)

,country
title,
Race Highlights | 2022 British Grand Prix,17


**Race Highlights | 2022 British Grand Prix**

# Часть 2. Визуализация данных [2 балла], кросс-проверка

**Matplotlib** - основная библиотека визуализации двумерных данных на языке python.

Рисунок создается с помощью последовательного вызова функций из модуля pyplot
(создание рисунка, задание области рисования, построение графиков в
области рисования, украшение графика метками). При этом различные
графические элементы рисунка (линии, подписи и т.д.) перекрывают
предыдущие, если они имеют общую площадь.

Основные типы графиков:

- **line plot** — обычный поточечный график, которым можно изображать кривые, соединяющие последовательные точки;
- **histogram** — гистограмма, показывающая распределение некоторое величины;
- **scatter plot** — диаграмма рассеивания, показывающая взаимосвязь двух величин;
- **bar plot** — столбчатая диаграмма, показывающий взаимосвязь количественной величины от категориальной.

### Оформление графиков

Важно, чтобы все графики выглядели опрятно и соответствовали следующим правилам:

🔷 Максимально поясняйте, что изображено на графике

```
✅ Добавляйте название (plt.title)
✅ Подписывайте оси (plt.xlabel, plt.ylabel)
✅ Подписывайте единицы измерения на осях
✅ Делайте понятные подписи
```

🔷 Следите за визуальной составляющей графика

```
✅ Подписи нигде не должны заезжать друг на друга
✅ Все линии на графиках должны быть чётко видны
❌ Не используйте похожие цвета или цвета, сливающиеся с фоном
✅ Используйте векторный (pdf/svg) формат при возможности
✅ Ограничивайте диапазоны значений там, где это возможно (например, вероятности можно ограничить диапазоном [0,1])
✅ Для улучшения демонстрации различий можно логарифмировать оси x и/или y (plt.xscale("log"), plt.yscale("log"))
❌ Не рекомендуется использовать 3d графики в презентациях: при возможности изобразите необходимое при помощи нескольких 2d графиков
```

🔷 Учитывайте размеры и вместимость графиков

```
✅ Графики должны быть комфортного для показа размера: все, что необходимо показать, должно быть хорошо видно
✅ Иногда можно разбивать логически график на несколько полотен, если на одно все не умещается
```

Больше примеров Вы можете найти в документации .

In [ ]:
import matplotlib
import matplotlib.pyplot as plt

#отображение графиков в ноутбуке
%matplotlib inline

#для четкой прорисовки графиков - осторожно!!! Может съедать много памяти
#закоменьте эту строку, если это случилось
%config InlineBackend.figure_format = 'svg'

Далее вам нужно построить несколько визуализаций табличных данных о видео, с которыми мы работали в 1 части задания.

**Внимание!**

- Проверьте, что при конвертации в html формат все графики остаются видны.
Это нужно для дальнейшей проверки в проверяющей системе. Проверить это
можно непосредственно загрузив задание в систему и посмотреть на свою
отправленную посылку (как это сделать - можно почитать в туториале о
проверяющей системе)
- При решении заданий вы можете воспользоваться любыми средствами для визуализации.
- Все графики должны выглядеть опрятно и соответствовать правилам, перечисленным в начале Части 2. В случае несоответствия хотя бы 3 пунктам оценка может быть снижена

### **Задание 1 [0.5 баллa]**

Постройте график (line plot) зависимости кол-ва выпущенных роликов от даты публикации.

**Подсказка**
Не нужно учитывать время публикации, только дату с точностью до дня.

Также обратите внимание, что один и тот же ролик может быть
опубликован в нескольких странах, и ему может соответствовать несколько
строк в таблице. На графике необходимо учесть каждый ролик только 1 раз.

Сначала может понадобиться перевести колонку даты публикации к типу datetime :

In [ ]:
df_new = df.drop_duplicates(subset=['video_id']).copy()

df_new.loc[:, 'publishedAt'] = pd.to_datetime(df_new['publishedAt']).dt.date

In [ ]:
x = df_new['publishedAt']
y = x.value_counts().sort_index()

In [ ]:
import matplotlib.pyplot as plt # или другая любимая библиотека визуализации
import matplotlib.dates as mdates

plt.plot(y.index, y.values)
plt.xticks([y.index[0], y.index[-1]])
plt.title("Зависимость кол-ва выпущенных роликов от даты публикации.")
plt.xlabel('Дата публикации')
plt.ylabel('Кол-во выпущенных роликов')

Text(0, 0.5, 'Кол-во выпущенных роликов')


?xml version="1.0" encoding="utf-8" standalone="no"?
 
 
 
 
 
 2024-10-09T14:47:16.788265 
 image/svg+xml 
 
 
 Matplotlib v3.7.1, https://matplotlib.org/ 
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 2022-06-04 -->
 
 
 
 
 
<path d="M 2113 2584 
Q 1688 2584 1439 2293 
Q 1191 2003 1191 1497 
Q 1191 994 1439 701 
Q 1688 409 2113 409 
Q 2538 409 2786 701 
Q 3034 994 3034 1497 
Q 3034 2003 2786 2293 
Q 2538 2584 2113 2584 
z
M 3366 4563 
L 3366 3988 
Q 3128 4100 2886 4159 
Q 2644 4219 2406 4219 
Q 1781 4219 1451 3797 
Q 1122 3375 1075 2522 
Q 1259 2794 1537 2939 
Q 1816 3084 2150 3084 
Q 2853 3084 3261 2657 
Q 3669 2231 3669 1497 
Q 3669 778 3244 343 
Q 2819 -91 2113 -91 
Q 1303 -91 875 529 
Q 447 1150 447 2328 
Q 447 3434 972 4092 
Q 1497 4750 2381 4750 
Q 2619 4750 2861 4703 
Q 3103 4656 3366 4563 
z
" id="DejaVuSans-36" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 2022-07-10 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- Дата публикации -->
 
 
 
<path d="M 2194 1759 
Q 1497 1759 1228 1600 
Q 959 1441 959 1056 
Q 959 750 1161 570 
Q 1363 391 1709 391 
Q 2188 391 2477 730 
Q 2766 1069 2766 1631 
L 2766 1759 
L 2194 1759 
z
M 3341 1997 
L 3341 0 
L 2766 0 
L 2766 531 
Q 2569 213 2275 61 
Q 1981 -91 1556 -91 
Q 1019 -91 701 211 
Q 384 513 384 1019 
Q 384 1609 779 1909 
Q 1175 2209 1959 2209 
L 2766 2209 
L 2766 2266 
Q 2766 2663 2505 2880 
Q 2244 3097 1772 3097 
Q 1472 3097 1187 3025 
Q 903 2953 641 2809 
L 641 3341 
Q 956 3463 1253 3523 
Q 1550 3584 1831 3584 
Q 2591 3584 2966 3190 
Q 3341 2797 3341 1997 
z
" id="DejaVuSans-430" transform="scale(0.015625)"> 
 
 
 
 
<path d="M 1991 3584 
Q 2741 3584 3169 3096 
Q 3597 2609 3597 1747 
Q 3597 888 3169 398 
Q 2741 -91 1991 -91 
Q 1238 -91 822 384 
Q 406 859 384 1747 
L 366 2463 
Q 350 2834 350 2963 
Q 350 3463 531 3831 
Q 816 4409 1308 4659 
Q 1800 4909 2625 4916 
Q 3006 4919 3175 4972 
L 3334 4516 
Q 3181 4453 3041 4450 
L 2259 4397 
Q 1866 4369 1628 4231 
Q 1006 3866 947 3388 
L 925 3213 
Q 1331 3584 1991 3584 
z
M 1991 3097 
Q 1528 3097 1259 2736 
Q 991 2375 991 1747 
Q 991 1119 1258 758 
Q 1525 397 1991 397 
Q 2450 397 2719 759 
Q 2988 1122 2988 1747 
Q 2988 2369 2719 2733 
Q 2450 3097 1991 3097 
z
" id="DejaVuSans-431" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0 -->
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 25 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 50 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 75 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 100 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 125 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 150 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 175 -->
 
 
 
 
 
 
 
 
<!-- Кол-во выпущенных роликов -->
 
 
 
 
<path d="M 1156 1613 
L 1156 459 
L 1975 459 
Q 2369 459 2575 607 
Q 2781 756 2781 1038 
Q 2781 1319 2575 1466 
Q 2369 1613 1975 1613 
L 1156 1613 
z
M 1156 3041 
L 1156 2072 
L 1913 2072 
Q 2238 2072 2444 2201 
Q 2650 2331 2650 2563 
Q 2650 2794 2444 2917 
Q 2238 3041 1913 3041 
L 1156 3041 
z
M 581 3500 
L 1950 3500 
Q 2566 3500 2897 3275 
Q 3228 3050 3228 2634 
Q 3228 2313 3059 2123 
Q 2891 1934 2559 1888 
Q 2956 1813 3175 1575 
Q 3394 1338 3394 981 
Q 3394 513 3033 256 
Q 2672 0 2003 0 
L 581 0 
L 581 3500 
z
" id="DejaVuSans-432" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<path clip-path="url(#pe1146311e8)" d="M 63.198352 276.5385 
L 72.216534 276.5385 
L 81.234716 273.675541 
L 90.252898 273.675541 
L 99.27108 269.381104 
L 108.289261 270.812583 
L 117.307443 265.086666 
L 126.325625 272.244062 
L 135.343807 275.107021 
L 144.361989 266.518145 
L 153.38017 256.49779 
L 162.398352 260.792228 
L 171.416534 242.182997 
L 180.434716 257.929269 
L 189.452898 262.223707 
L 198.47108 269.381104 
L 207.489261 260.792228 
L 216.507443 240.751518 
L 225.525625 250.771873 
L 234.543807 242.182997 
L 243.561989 250.771873 
L 252.58017 239.320038 
L 261.598352 237.888559 
L 270.616534 235.025601 
L 279.634716 223.573766 
L 2

### **Задание 2 [0.5 баллa]**

Постройте диаграмму (bar plot) кол-ва опубликованных роликов в зависимости от страны.

In [ ]:
df_new = df.groupby('country')['video_id'].nunique().sort_values(ascending=False)
df_new.plot(kind='bar')
plt.title("Кол-во опубликованных роликов в зависимости от страны")
plt.xlabel('Страна')
plt.ylabel('Кол-во выпущенных роликов')

Text(0, 0.5, 'Кол-во выпущенных роликов')


?xml version="1.0" encoding="utf-8" standalone="no"?
 
 
 
 
 
 2024-10-09T14:50:58.043712 
 image/svg+xml 
 
 
 Matplotlib v3.7.1, https://matplotlib.org/ 
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- EG -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- LY -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- LB -->
 
 
<path d="M 1259 2228 
L 1259 519 
L 2272 519 
Q 2781 519 3026 730 
Q 3272 941 3272 1375 
Q 3272 1813 3026 2020 
Q 2781 2228 2272 2228 
L 1259 2228 
z
M 1259 4147 
L 1259 2741 
L 2194 2741 
Q 2656 2741 2882 2914 
Q 3109 3088 3109 3444 
Q 3109 3797 2882 3972 
Q 2656 4147 2194 4147 
L 1259 4147 
z
M 628 4666 
L 2241 4666 
Q 2963 4666 3353 4366 
Q 3744 4066 3744 3513 
Q 3744 3084 3544 2831 
Q 3344 2578 2956 2516 
Q 3422 2416 3680 2098 
Q 3938 1781 3938 1306 
Q 3938 681 3513 340 
Q 3088 0 2303 0 
L 628 0 
L 628 4666 
z
" id="DejaVuSans-42" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- TN -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- DZ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- MA -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- FR -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- CY -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- SI -->
 
 
<path d="M 3425 4513 
L 3425 3897 
Q 3066 4069 2747 4153 
Q 2428 4238 2131 4238 
Q 1616 4238 1336 4038 
Q 1056 3838 1056 3469 
Q 1056 3159 1242 3001 
Q 1428 2844 1947 2747 
L 2328 2669 
Q 3034 2534 3370 2195 
Q 3706 1856 3706 1288 
Q 3706 609 3251 259 
Q 2797 -91 1919 -91 
Q 1588 -91 1214 -16 
Q 841 59 441 206 
L 441 856 
Q 825 641 1194 531 
Q 1563 422 1919 422 
Q 2459 422 2753 634 
Q 3047 847 3047 1241 
Q 3047 1584 2836 1778 
Q 2625 1972 2144 2069 
L 1759 2144 
Q 1053 2284 737 2584 
Q 422 2884 422 3419 
Q 422 4038 858 4394 
Q 1294 4750 2059 4750 
Q 2388 4750 2728 4690 
Q 3069 4631 3425 4513 
z
" id="DejaVuSans-53" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- IL -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- TR -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- HR -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- ES -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- GR -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- MT -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- IT -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- AL -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- BA -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- ME -->
 
 
 
 
 
 
 
<!-- Страна -->
 
 
 
 
 
<path d="M 2194 1759 
Q 1497 1759 1228 1600 
Q 959 1441 959 1056 
Q 959 750 1161 570 
Q 1363 391 1709 391 
Q 2188 391 2477 730 
Q 2766 1069 2766 1631 
L 2766 1759 
L 2194 1759 
z
M 3341 1997 
L 3341 0 
L 2766 0 
L 2766 531 
Q 2569 213 2275 61 
Q 1981 -91 1556 -91 
Q 1019 -91 701 211 
Q 384 513 384 1019 
Q 384 1609 779 1909 
Q 1175 2209 1959 2209 
L 2766 2209 
L 2766 2266 
Q 2766 2663 2505 2880 
Q 2244 3097 1772 3097 
Q 1472 3097 1187 3025 
Q 903 2953 641 2809 
L 641 3341 
Q 956 3463 1253 3523 
Q 1550 3584 1831 3584 
Q 2591 3584 2966 3190 
Q 3341 2797 3341 1997 
z
" id="DejaVuSans-430" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 0 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 25 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 50 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 75 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 100 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 125 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 150 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 175 -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- 200 -->
 
 
 
 
 
 
 
 
<!-- Кол-во выпущенных роликов -->
 
 
 
 
 
 
<path d="M 1156 1613 
L 1156 459 
L 1975 459 
Q 2369 459 2575 607 
Q 2781 756 2781 1038 
Q 2781 1319 2575 1466 
Q 2369 1613 1975 1613 
L 1156 1613 
z
M 1156 3041 
L 1156 2072 
L 1913 2072 
Q 2238 2072 2444 2201 
Q 2650 2331 2650 2563 
Q 2650 2794 2444 2917 
Q 2238 3041 1913 3041 
L 1156 3041 
z
M 581 3500 
L 1950 3500 
Q 2566 3500 2897 3275 
Q 3228 3050 3228 2634 
Q 3228 2313 3059 2123 
Q 2891 1934 2559 1888 
Q 2956 1813 3175 1575 
Q 3394 1338 3394 981 
Q 3394 513 3033 256 
Q 2672 0 2003 0 
L 581 0 
L 581 3500 
z
" id="DejaVuSans-432" transform="scale(0.01562

### **Задание 3 [1 балл]**

Постройте график (scatter plot)
зависимости кол-ва комментариев для ролика от кол-ва просмотров для
роликов во Франции (country='FR'). На графике не нужно учитывать ролики с
отключенными комментариями (comments_disabled=True).

Какой характер зависмости получился?

In [ ]:
categories = ['view_count', 'comment_count']
df_new = df[(df['country'] == 'FR') & (df['comments_disabled'] == False)][categories]
x = df_new['comment_count'].tolist()
y = df_new['view_count'].tolist()
plt.scatter(x, y, s = 10)
plt.xscale('log')
plt.yscale('log')
plt.title("Зависимость кол-ва комментариев от просмотров")
plt.xlabel('Кол-во комментариев')
plt.ylabel('Кол-во просмотров')

Text(0, 0.5, 'Кол-во просмотров')


?xml version="1.0" encoding="utf-8" standalone="no"?
 
 
 
 
 
 2024-10-09T15:18:53.138813 
 image/svg+xml 
 
 
 Matplotlib v3.7.1, https://matplotlib.org/ 
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{2}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{3}}$ -->
 
 
<path d="M 2597 2516 
Q 3050 2419 3304 2112 
Q 3559 1806 3559 1356 
Q 3559 666 3084 287 
Q 2609 -91 1734 -91 
Q 1441 -91 1130 -33 
Q 819 25 488 141 
L 488 750 
Q 750 597 1062 519 
Q 1375 441 1716 441 
Q 2309 441 2620 675 
Q 2931 909 2931 1356 
Q 2931 1769 2642 2001 
Q 2353 2234 1838 2234 
L 1294 2234 
L 1294 2753 
L 1863 2753 
Q 2328 2753 2575 2939 
Q 2822 3125 2822 3475 
Q 2822 3834 2567 4026 
Q 2313 4219 1838 4219 
Q 1578 4219 1281 4162 
Q 984 4106 628 3988 
L 628 4550 
Q 988 4650 1302 4700 
Q 1616 4750 1894 4750 
Q 2613 4750 3031 4423 
Q 3450 4097 3450 3541 
Q 3450 3153 3228 2886 
Q 3006 2619 2597 2516 
z
" id="DejaVuSans-33" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{4}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{5}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- Кол-во комментариев -->
 
 
 
 
 
 
<path d="M 1156 1613 
L 1156 459 
L 1975 459 
Q 2369 459 2575 607 
Q 2781 756 2781 1038 
Q 2781 1319 2575 1466 
Q 2369 1613 1975 1613 
L 1156 1613 
z
M 1156 3041 
L 1156 2072 
L 1913 2072 
Q 2238 2072 2444 2201 
Q 2650 2331 2650 2563 
Q 2650 2794 2444 2917 
Q 2238 3041 1913 3041 
L 1156 3041 
z
M 581 3500 
L 1950 3500 
Q 2566 3500 2897 3275 
Q 3228 3050 3228 2634 
Q 3228 2313 3059 2123 
Q 2891 1934 2559 1888 
Q 2956 1813 3175 1575 
Q 3394 1338 3394 981 
Q 3394 513 3033 256 
Q 2672 0 2003 0 
L 581 0 
L 581 3500 
z
" id="DejaVuSans-432" transform="scale(0.015625)"> 
 
 
 
 
 
 
<path d="M 2194 1759 
Q 1497 1759 1228 1600 
Q 959 1441 959 1056 
Q 959 750 1161 570 
Q 1363 391 1709 391 
Q 2188 391 2477 730 
Q 2766 1069 2766 1631 
L 2766 1759 
L 2194 1759 
z
M 3341 1997 
L 3341 0 
L 2766 0 
L 2766 531 
Q 2569 213 2275 61 
Q 1981 -91 1556 -91 
Q 1019 -91 701 211 
Q 384 513 384 1019 
Q 384 1609 779 1909 
Q 1175 2209 1959 2209 
L 2766 2209 
L 2766 2266 
Q 2766 2663 2505 2880 
Q 2244 3097 1772 3097 
Q 1472 3097 1187 3025 
Q 903 2953 641 2809 
L 641 3341 
Q 956 3463 1253 3523 
Q 1550 3584 1831 3584 
Q 2591 3584 2966 3190 
Q 3341 2797 3341 1997 
z
" id="DejaVuSans-430" transform="scale(0.015625)"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{5}}$ -->
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<!-- $\mathdefault{10^{6}}$ -->
 
 
<path d="M 2113 2584 
Q 1688 2584 1439 2293 
Q 1191 2003 1191 1497 
Q 1191 994 1439 701 
Q 1688 409 2113 409 
Q 2538 409 2786 701 
Q 3034 994 3034 1497 
Q 3034 2003 2786 2293 
Q 2538 2584 2113 2584 
z
M 3366 4563 
L 3366 3988 
Q 3128 4100 2886 4159 
Q 2644 4219 2406 4219 
Q 1781 4219 1451 3797 
Q 1122 3375 1075 2522 
Q 1259 2794 1537 2939 
Q 1816 3084 2150 3084 
Q 2853 3084 3261 2657 
Q 3669 2231 3669 1497 
Q 3669 778 3244 343 
Q 2819 -91 2113 -91 
Q 1303 -91 875 529 
Q 447 1150 447 2328 
Q 447 3434 972 4092 
Q 1497 4750 2381 4750 
Q 2619 4750 2861 4703 
Q 3103 4656 3366 456

**Ваш ответ тут:**

# Часть 3. Numpy [3 балла], сдача в проверяющую систему

In [ ]:
import numpy as np

**Внимание!**

- Файлы base_functions.py и base_functions_vectorised.py с реализованными функциями необходимо сдать на тестирование в проверяющей системе.

### **Задачи 1-3 [3 баллa]**

Для каждой из задач ниже привидите 2 реализации: одна без использования numpy, а вторая полностью векторизованная ( без использования циклов/map/list comprehension ).
Реализации без использования векторизации нужно записать в файл
base_functions.py, а векторизованные — в файл
base_functions_vectorised.py (см. шаблоны).

**Внимание!**

Векторизованные реализации с использованием циклов/map/list comprehension оцениваются в 0 баллов .

```
*   Невекторизованные варианты оцениваются - по 0.3 балла
*   Векторизованные варианты - по 0.7 балла
```

#### **Задача 1 [0.3 + 0.7 баллов]**

Вам
дан двумерный непустой массив X размера n x m вещественных чисел.
Создайте на его основе массив состоящий: из каждого 4го (начиная с 0го и
до конца) элемента по оси размерности n (axis=0), начиная с 120 до 500
(не включитально) с шагом 5 элементов по оси размерности m (axis=1).

Гарантируется, что m >= 500

\ Пример:

```
  Input: X = [[1, 2, 3, ..., 500], [1, 2, 3, ..., 500], [1, 2, 3, ..., 500], [1, 2, 3, ..., 500], [1, 2, 3, ..., 500]]
  Output: [[121, 126, 131, ..., 496]]
```

#### **Задача 2 [0.3 + 0.7 баллов]**

Вам дан двумерный непустой массив X целых чисел размера n x m.

Посчитайте сумму неотрицательных элементов на главной диагонали
прямоугольной матрицы X. Если неотрицательных элементов на диагонали
нет, то верните -1.

Пример:

```
  Input: X = np.array([[-1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
  Output: 3
```

#### **Задача 1 [0.3 + 0.7 баллов]**

Дан
двумерный непустой массив X вещественных чисел размера n x m. Считая
для каждого столбца отдельно, замените в копии массива те значения,
которые больше 1.5 средних по столбцу или меньше 1/4 среднего по столбцу
на число -1

**Подсказка**

Возможно пригодится функция deepcopy из модуля copy

# **Задание [Bonus][0.5 балла]** , кросс-проверка

Вставьте/придумайте мем про pandas/машинное обучение/Python или просто милую картинку с пандой

**Внимание!** самый простой способ вставить картинку будет через Google Colab (даже если вы изначально делали не в нем). Нажмите на "+ Text" , в появившейся ячейке сделайте прикрепление картинки (как на скринах). Тогда ваша картинка "зашифруется" и будет корректно отображаться при конвертации в html

**Внимание!** Если вы хотите вставить
картинку в других системах работы с ipynb, то можете просто скопировать
картинку в буфер обмена и вставить содержимое в текстовую ячейку (ctrl-c
ctrl-v)

Ваша картинка тут